xtokens/src/lib.rs

```diff
@@ -467,43 +469,140 @@ pub mod module {
+				if (fee != *asset && non_fee_reserve.is_none()) || asset_len == 1 {
+					non_fee_reserve = asset.reserve();
+				}
+				// make sure all non fee assets share the same reserve
+				if non_fee_reserve.is_some() {
+					ensure!(
+						non_fee_reserve == asset.reserve(),
+						Error::<T>::DistinctReserveForAssetAndFee
+					);
+				}
 			}
 
-			let (transfer_kind, dest, reserve, recipient) = Self::transfer_kind(&fee, &dest)?;
-			let mut msg = match transfer_kind {
-				SelfReserveAsset => Self::transfer_self_reserve_asset(
-					assets.clone(),
-					fee.clone(),
-					dest.clone(),
-					recipient,
-					dest_weight,
-				)?,
-				ToReserve => {
-					Self::transfer_to_reserve(assets.clone(), fee.clone(), dest.clone(), recipient, dest_weight)?
+			// In case that fee reserve != asset reserve, there are two xcm sent from
+			// sender. first xcm sent to fee reserve chain which route to dest chain. second
+			// xcm directly sent to dest chain. the fee amount in fee asset is split into
+			// two parts. 1. fee asset sent to fee reserve chain = fee_amount - min_xcm_fee
+			// 2. fee asset sent to dest reserve chain = min_xcm_fee
+			// the first part + second part = fee amount in fee asset
+			let fee_reserve = fee.reserve();
+			if fee_reserve != non_fee_reserve {
+				// Current only support `ToReserve` with relay-chain asset as fee. other case
+				// like `NonReserve` or `SelfReserve` with relay-chain fee is not support.
+				ensure!(non_fee_reserve == dest.chain_part(), Error::<T>::InvalidAsset);
+
+				let reserve_location = non_fee_reserve.clone().ok_or(Error::<T>::AssetHasNoReserve)?;
+				let min_xcm_fee = T::MinXcmFee::get(&reserve_location);
+
+				// min xcm fee should less than user fee
+				let fee_to_dest = reset_fee(&fee, min_xcm_fee);
+				ensure!(fee_to_dest < fee, Error::<T>::InvalidAsset);
+
+				let mut assets_to_dest = MultiAssets::new();
+				for i in 0..asset_len {
+					let asset = assets.get(i).ok_or(Error::<T>::AssetIndexNonExistent)?;
+					if fee != *asset {
+						assets_to_dest.push(asset.clone());
+					} else {
+						assets_to_dest.push(fee_to_dest.clone());
+					}
 				}
-				ToNonReserve => Self::transfer_to_non_reserve(
+
+				let mut assets_to_fee_reserve = MultiAssets::new();
+				let asset_to_fee_reserve = subtract_fee(&fee, min_xcm_fee);
+				assets_to_fee_reserve.push(asset_to_fee_reserve.clone());
+
+				// First xcm sent to fee reserve chain and routed to dest chain.
+				// The `SelfLocation` current is (1, Parachain(id)) refer to sender parachain.
+				// we use `SelfLocation` to fund fee to sender's parachain sovereign account on
+				// destination chain, which asset is originated from sender account on sender
+				// chain. This means if user setup too much fee, the fee is not returned to
+				// user, instead deposit to sibling parachain sovereign account on dest chain.
+				// Notice: if parachain set `SelfLocation` to (0, Here), it'll be error!
```

Comment:
This should be documented in `SelfLocation`.

Reply:
move to README